File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_py/code_output/DecryptionUtils.py
class DecryptionUtils:
    def __init__(self, key):
        self.key = key

    def caesarDecipher(self, ciphertext, shift):
        plaintext = []
        shift = shift % 26
        for ch in ciphertext:
            if ch.isalpha():
                ascii_offset = ord('A') if ch.isupper() else ord('a')
                shifted_char = chr((ord(ch) - ascii_offset - shift + 26) % 26 + ascii_offset)
                plaintext.append(shifted_char)
            else:
                plaintext.append(ch)
        return ''.join(plaintext)

    def vigenereDecipher(self, ciphertext):
        decrypted_text = []
        key_index = 0
        for ch in ciphertext:
            if ch.isalpha():
                shift = ord(self.key[key_index % len(self.key)].lower()) - ord('a')
                base = ord('A') if ch.isupper() else ord('a')
                decrypted_char = chr((ord(ch) - base - shift + 26) % 26 + base)
                decrypted_text.append(decrypted_char)
                key_index += 1
            else:
                decrypted_text.append(ch)
        return ''.join(decrypted_text)

    def railFenceDecipher(self, encryptedText, rails):
        fence = [['\0'] * len(encryptedText) for _ in range(rails)]

        direction = -1
        row = 0
        col = 0

        for i in range(len(encryptedText)):
            if row == 0 or row == rails - 1:
                direction = -direction
            fence[row][col] = '*'
            col += 1
            row += direction

        index = 0
        for i in range(rails):
            for j in range(len(encryptedText)):
                if fence[i][j] == '*':
                    fence[i][j] = encryptedText[index]
                    index += 1

        plain_text = []
        direction = -1
        row = 0
        col = 0

        for i in range(len(encryptedText)):
            if row == 0 or row == rails - 1:
                direction = -direction
            plain_text.append(fence[row][col])
            col += 1
            row += direction

        return ''.join(plain_text)

File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_cpp/code_output/DecryptionUtils.cpp
#include <string>
#include <cctype>
#include <vector>
#include <cstring>

class DecryptionUtils {
private:
    std::string key;

public:
    DecryptionUtils(const std::string& key) : key(key) {}

    std::string caesarDecipher(const std::string& ciphertext, int shift) {
        std::string plaintext;
        shift = shift % 26;
        for (char ch : ciphertext) {
            if (std::isalpha(ch)) {
                int asciiOffset = std::isupper(ch) ? 'A' : 'a';
                char shiftedChar = static_cast<char>((ch - asciiOffset - shift + 26) % 26 + asciiOffset);
                plaintext += shiftedChar;
            } else {
                plaintext += ch;
            }
        }
        return plaintext;
    }

    std::string vigenereDecipher(const std::string& ciphertext) {
        std::string decryptedText;
        int keyIndex = 0;
        for (char ch : ciphertext) {
            if (std::isalpha(ch)) {
                int shift = std::tolower(key[keyIndex % key.length()]) - 'a';
                char base = std::isupper(ch) ? 'A' : 'a';
                char decryptedChar = static_cast<char>((ch - base - shift + 26) % 26 + base);
                decryptedText += decryptedChar;
                keyIndex++;
            } else {
                decryptedText += ch;
            }
        }
        return decryptedText;
    }

    std::string railFenceDecipher(const std::string& encryptedText, int rails) {
        int len = encryptedText.length();
        std::vector<std::vector<char>> fence(rails, std::vector<char>(len, '\0'));

        int direction = -1;
        int row = 0, col = 0;

        for (int i = 0; i < len; ++i) {
            if (row == 0 || row == rails - 1) {
                direction *= -1;
            }
            fence[row][col++] = '*';
            row += direction;
        }

        int index = 0;
        for (int i = 0; i < rails; ++i) {
            for (int j = 0; j < len; ++j) {
                if (fence[i][j] == '*') {
                    fence[i][j] = encryptedText[index++];
                }
            }
        }

        std::string plainText;
        direction = -1;
        row = 0;
        col = 0;

        for (int i = 0; i < len; ++i) {
            if (row == 0 || row == rails - 1) {
                direction *= -1;
            }
            plainText += fence[row][col++];
            row += direction;
        }

        return plainText;
    }
};